<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [8]:
# текст запроса
query_3_1 = f'''select DISTINCT id
           from public.vacancies
        '''
df = pd.read_sql_query(query_3_1, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\466410155.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [10]:
query_3_2 = f'''select DISTINCT id
           from public.employers
        '''
df = pd.read_sql_query(query_3_2, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\3869639284.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [11]:
query_3_3 = f'''select *
           from public.areas
        '''
df = pd.read_sql_query(query_3_3, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\2631811427.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [15]:
query_3_4 = f'''select *
           from public.industries
        '''
df = pd.read_sql_query(query_3_4, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\1824885113.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


294

***

В данных представлены вакансии, которые охватывают тяжелую и лёгкую промышленности, пишевую индустрию и сферу услуг. Вакансии представлены по всему миру, но большинство в России, которая особенно подробно разбита на регионы и города.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [23]:
query_4_1 = f'''select a.name, count(v.id) cnt
           from public.vacancies v join public.areas a on v.area_id = a.id
           group by a.name
           order by cnt  desc
        '''
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\2994418298.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [25]:
query_4_2 = f'''select count(*), vacancies.salary_from, vacancies.salary_to, vacancies.salary_from + vacancies.salary_to
           from public.vacancies
           where vacancies.salary_to is not null and vacancies.salary_from is null
           or vacancies.salary_to is null and vacancies.salary_from is not null
           or vacancies.salary_to is not null and vacancies.salary_from is not null
           group by vacancies.id
        '''
df = pd.read_sql_query(query_4_2, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\1191996193.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


24073

In [8]:
query_4_2 = f'''select *
           from public.vacancies
           where vacancies.salary_to IS null
        '''
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\nikol\AppData\Local\Temp\ipykernel_18172\1447232747.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


id                                               name  \
0      55312386                                Компьютерный Мастер   
1      55843014                            Системный администратор   
2      54525964                      Lead Java Developer to Poland   
3      54525965                      Lead Java Developer to Poland   
4      55354053  Специалист службы поддержки с техническими зна...   
...         ...                                                ...   
36502  55904166                                   Менеджер проекта   
36503  55589514           Специалист по информационным технологиям   
36504  55029668                     Руководитель процессного офиса   
36505  66172263                         Начальник сервисной службы   
36506  55547842                                   DataOps Engineer   

                                              key_skills          schedule  \
0      Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1      Средства криптографической защиты информации\t...       Полный день   
2        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                                   None  Удаленная работа   
...                                                  ...               ...   
36502  Управление проектами\tДеловая коммуникация\tОр...       Полный день   
36503  Электронный документооборот\t1С: Предприятие 8...       Полный день   
36504  Оптимизация бизнес-процессов\tСтратегическое м...       Полный день   
36505  Организаторские навыки\tУправление персоналом\...       Полный день   
36506                                               None  Удаленная работа   

               experience           employment  salary_from salary_to  \
0               Нет опыта     Полная занятость      64000.0      None   
1           От 3 до 6 лет     Полная занятость          NaN      None   
2           От 3 до 6 лет     Полная занятость          NaN      None   
3           От 3 до 6 лет     Полная занятость          NaN      None   
4               Нет опыта  Частичная занятость      15000.0      None   
...                   ...                  ...          ...       ...   
36502       От 3 до 6 лет     Полная занятость          NaN      None   
36503       От 3 до 6 лет     Полная занятость          NaN      None   
36504       От 3 до 6 лет     Полная занятость          NaN      None   
36505  От 1 года до 3 лет     Полная занятость      60000.0      None   
36506       От 3 до 6 лет     Полная занятость          NaN      None   

       area_id  employer_id  
0         1516      5724811  
1          130      4903713  
2          160        69961  
3          159        69961  
4         1955         1740  
...        ...          ...  
36502       53       895094  
36503       66      5109332  
36504       26        51333  
36505     1930      4094700  
36506      160      4978627  

[36507 rows x 10 columns]

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [28]:
print("salary_from:", df["salary_from"].mean())
print("salary_to:", df["salary_to"].mean())

salary_from: 71064.65790086107
salary_to: 110536.74192277384


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [29]:
query_4_4 = f'''select vacancies.schedule, vacancies.employment, count(id)
           from public.vacancies
           group by vacancies.schedule, vacancies.employment
           order by count DESC
        '''
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\1400374966.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [30]:
query = f'''select *
           from public.vacancies
        '''
df = pd.read_sql_query(query, connection)
df["experience"].value_counts()

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\2995116292.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


От 1 года до 3 лет    26152
От 3 до 6 лет         14511
Нет опыта              7197
Более 6 лет            1337
Name: experience, dtype: int64

***

Итак, в данных есть вакансии в 1362 региона, однако почти половина вакансий находится в десяти крупнейших городах России. Лишь половина вакансий чётко обозначают границы зарплат, что, видимо, говорит о том, что условия могут измениться в зависимости от соискателя. В среднем предлагаемые около 100000 Р, но стоит учитывать, что семь из десяти вакансий предлагаются людям либо с малым опытом работы, либо без опыта вовсе, хоть обычно и на полный рабочий день.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [49]:
query_5_1 = f'''select e.name, count(*) number_of_vac
                from public.employers e join public.vacancies v on e.id = v.employer_id
                group by e.name
                order by count(*) DESC
                limit 5'''
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\nikol\AppData\Local\Temp\ipykernel_31284\3913119699.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  number_of_vac
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [6]:
query_5_2 = f'''select a.name, count(DISTINCT e.id) e_count, count(DISTINCT v.id) v_count
           from public.areas a left join public.employers e on a.id = e.area
           left join public.vacancies v on v.area_id = a.id
           where v.id is null
           group by a.name
           order by e_count DESC, v_count DESC
        '''
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\nikol\AppData\Local\Temp\ipykernel_19568\1309142211.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  e_count  v_count
0                            Россия      410        0
1                         Казахстан      207        0
2                Московская область       75        0
3                Краснодарский край       19        0
4                          Беларусь       18        0
..                              ...      ...      ...
588                        Кувандык        0        0
589                        Куйбышев        0        0
590  Куйбышево (Ростовская область)        0        0
591                        Култаево        0        0
592                        Кульсары        0        0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [7]:
query_5_3 = f'''select e.name, count(DISTINCT v.area_id) number_of_regions
           from public.employers e left join public.vacancies v on e.id = v.employer_id
           group by e.name
           order by count(DISTINCT v.area_id) DESC
        '''
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\nikol\AppData\Local\Temp\ipykernel_19568\3923150816.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  number_of_regions
0                      Яндекс                181
1                  Ростелеком                152
2                  Спецремонт                116
3      Поляков Денис Иванович                 88
4                   ООО ЕФИН                  71
...                       ...                ...
23170                  СДЕЛКА                  0
23171            Alandr Group                  0
23172                     СДК                  0
23173      Lemon Land Lombard                  0
23174                  Махина                  0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [12]:
query_5_4 = f'''select *
           from public.employers_industries ei left join public.industries i on i.id = ei.industry_id
           right join public.employers e on e.id = ei.employer_id
           where i.name is null
        '''
df = pd.read_sql_query(query_5_4, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_19568\3490707110.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [14]:
query_5_5 = f'''select e.name, count(i.name)
           from public.employers_industries ei left join public.industries i on i.id = ei.industry_id
           right join public.employers e on e.id = ei.employer_id
           group by e.name
           having count(i.name) = 4
           order by e.name
        '''
df = pd.read_sql_query(query_5_5, connection)
df["name"][2]

C:\Users\nikol\AppData\Local\Temp\ipykernel_19568\2984995779.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


'2ГИС'

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [19]:
query_5_6 = f'''select count(e.name) cnt
           from public.employers_industries ei left join public.industries i on i.id = ei.industry_id
           right join public.employers e on e.id = ei.employer_id
           group by i.name
           having i.name = 'Разработка программного обеспечения'
        '''
df = pd.read_sql_query(query_5_6, connection)
df['cnt'][0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_19568\3831114428.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [37]:
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

url
list_city = list(pd.read_html(url)[1]["Город"])
list_city[-3] = list_city[-3][:-3] #замена Воронеж[a] на Воронеж
list_city = tuple(list_city)

In [40]:
query_5_7 = f'''(select a.name, count(v.id)
           from public.vacancies v join public.areas a on v.area_id = a.id join public.employers e on e.id = v.employer_id
           where e.name = 'Яндекс' and a.name in {list_city}
           group by a.id)
           UNION ALL
           select 'Total', count(v.id)
           from public.vacancies v join public.areas a on v.area_id = a.id join public.employers e on e.id = v.employer_id
           where e.name = 'Яндекс' and a.name in {list_city}

            order by 2
        '''
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\nikol\AppData\Local\Temp\ipykernel_26684\3622026219.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7            Самара     26
8               Уфа     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

Наибольшее число вакансий предоставляют компании из сферы IT, коммуникации и из добывающей промышленности России. Однако примерно треть компаний не указывают свою сферу деятельности. Крупнейшая компания Яндекс предлагает работу в 181 регионе, причём 485 вакансии предлагаются в 16 городах миллионниках России.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [48]:
query_6_1 = f'''select count(DISTINCT id)
            from public.vacancies
            where LOWER(name) like '%данн%' or LOWER(name) like '%data%'
        '''
df = pd.read_sql_query(query_6_1, connection)
df["count"][0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_26684\1361836130.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [50]:
query_6_2 = f'''select *
            from public.vacancies
            where (LOWER(name) like '%data scientist%' or LOWER(name) like '%data science%' or LOWER(name) like '%исследователь данных%' or name like '%ML%' or LOWER(name) like '%machine learning%' or LOWER(name) like '%машинн%обучен%') and (LOWER(name) like '%junior%' or experience like '%Нет опыта%' or schedule like '%Стажировка%') and not (name like '%HTML%')
        '''
df = pd.read_sql_query(query_6_2, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_26684\2199138281.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [51]:
query = f'''select *
            from public.vacancies
            where (key_skills like '%SQL%' or key_skills like '%postgres%') and (LOWER(name) like '%data scientist%' or LOWER(name) like '%data science%' or LOWER(name) like '%исследователь данных%' or name like '%ML%' or LOWER(name) like '%machine learning%' or LOWER(name) like '%машинн%обучен%') and not (name like '%HTML%')
        '''
df = pd.read_sql_query(query, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_26684\2912045549.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [53]:
query_6_4 = f'''select *
            from public.vacancies
            where (LOWER(key_skills) like '%python%') and (LOWER(name) like '%data scientist%' or LOWER(name) like '%data science%' or LOWER(name) like '%исследователь данных%' or name like '%ML%' or LOWER(name) like '%machine learning%' or LOWER(name) like '%машинн%обучен%') and not (name like '%HTML%')
        '''
df = pd.read_sql_query(query_6_4, connection)
df.shape[0]

C:\Users\nikol\AppData\Local\Temp\ipykernel_26684\1077973632.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [55]:
query = f'''select key_skills
            from public.vacancies
            where (LOWER(name) like '%data scientist%' or LOWER(name) like '%data science%' or LOWER(name) like '%исследователь данных%' or name like '%ML%' or LOWER(name) like '%machine learning%' or LOWER(name) like '%машинн%обучен%') and not (name like '%HTML%') AND key_skills IS NOT NULL
        '''
df = pd.read_sql_query(query, connection)
round(df['key_skills'].apply(lambda x: len(x.split('\t'))).mean(),2)

C:\Users\nikol\AppData\Local\Temp\ipykernel_26684\3963169523.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [56]:
query = f'''select experience, AVG((coalesce(salary_from, salary_to, NULL) + coalesce(salary_to, salary_from, NULL))/2)
            from public.vacancies
            where (LOWER(name) like '%data scientist%' or LOWER(name) like '%data science%' or LOWER(name) like '%исследователь данных%' or name like '%ML%' or LOWER(name) like '%machine learning%' or LOWER(name) like '%машинн%обучен%') and not (name like '%HTML%') AND (salary_from IS NOT NULL or salary_to IS NOT NULL)
            group by experience
        '''
df = pd.read_sql_query(query, connection)
df

C:\Users\nikol\AppData\Local\Temp\ipykernel_26684\2557347930.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience            avg
0           Нет опыта   74642.857143
1  От 1 года до 3 лет  139674.750000
2       От 3 до 6 лет  243114.666667

***

Итак, для дата-сайнтиста есть 1771 вакансия, то есть меньше 4 процентов от общего числа вакансий в базе данных. Для начинающих только 0.01%. SQL или postgres требуется в качестве ключевого навыка в каждой десятой вакансии для ДС. Python требуется в каждом пятой вакансии. В среднем работадатель требует от ДС около 6-7 ключевых навыков с зарплатой 75000 Р для человека без опыта работы и 240000 Р для человека со стажем от 3 лет.

# Общий вывод по проекту

В данных представлены вакансии, которые охватывают тяжелую и лёгкую промышленности, пишевую индустрию и сферу услуг. Кроме программистов также представлены вакансии связанные с цифровой индустрией, например руководителей проектов или компьютерных мастеров. Вакансии представлены по всему миру, но большинство в России, которая особенно подробно разбита на регионы и города. В данных есть вакансии из 1362 регионов, однако почти половина вакансий находится в десяти крупнейших городах России. Москва выделяется на фоне остальных регионов количеством вакансий (в 2 раза больше, чем у второго города по количеству вакансий Санкт-Петербурга) и предлагаемой ЗП. Лишь половина вакансий чётко обозначают границы зарплат, что, видимо, говорит о том, что условия могут измениться в зависимости от соискателя. В среднем предлагаемые около 100000 Р, но стоит учитывать, что семь из десяти вакансий предлагаются людям либо с малым опытом работы, либо без опыта вовсе. На момент сохранения базы данных удаленный вид работы не был так распространен, поэтому и количество вакансий лишь около 4%, в подавляющем большинстве же предлагается полный рабочий день (видимо в офисе). Наибольшее число вакансий предоставляют компании из сферы IT, коммуникации и из добывающей промышленности России. Однако примерно треть компаний не указывают свою сферу деятельности. Крупнейшая компания Яндекс предлагает работу в 181 регионе, причём 485 вакансии предлагаются в 16 городах миллионниках России. Итак, для дата-сайнтиста есть 1771 вакансия, то есть меньше 4 процентов от общего числа вакансий в базе данных. Для начинающих только 0.01%. SQL или postgres требуется в качестве ключевого навыка в каждой десятой вакансии для ДС. Python требуется в каждом пятой вакансии. Часто требуется знание Pandas и Linux. В среднем работадатель требует от ДС около 6-7 ключевых навыков с зарплатой 75000 Р для человека без опыта работы и 240000 Р для человека со стажем от 3 лет.